seastar/src/clocks.rs

```diff
@@ -6,3 +8,135 @@ mod ffi {
 }
 
 use ffi::*;
+
+// Macro used to generate the implemenetation of the instant types.
+// - `InstantName` - name of the instant type,
+// - `Instant` - must be equal to `InstantName`,
+// - `Duration` - name of the corresponding duration type,
+// - `Clock` - name of the corresponding clock.
+macro_rules! instant_impl {
+    ($InstantName:ident, $Instant:ty, $Duration:ty, $Clock:ty) => {
+        #[doc = concat!("Type used by [`", stringify!($Clock), "`]")]
+        /// to represent points in time.
+        ///
+        #[doc = concat!("`", stringify!($Instant), "` ")]
+        /// is is implemented as if it stores a value of
+        #[doc = concat!(" [`", stringify!($Duration), "`] ")]
+        /// indicating the time interval from the start of the
+        #[doc = concat!(" [`", stringify!($Clock), "`]'s epoch.")]
+        #[derive(Eq, PartialEq, Ord, PartialOrd, Clone, Copy, Hash, Default, Debug)]
+        pub struct $InstantName {
+            pub(crate) nanos: i64,
+        }
+
+        impl $Instant {
+            /// Returns `Some(t)` where `t` is equal to `self + duration` if `t` is
+            /// inside the bounds of the underlying data structure, None otherwise.
+            pub fn checked_add(&self, duration: $Duration) -> Option<Self> {
+                match self.nanos.checked_add(duration.nanos) {
+                    Some(nanos) => Some(Self { nanos }),
+                    None => None,
+                }
+            }
+
+            /// Returns `Some(t)` where `t` ii equal to `self - duration` if `t` is
+            /// inside the bounds of the underlying data structure, None otherwise.
+            pub fn checked_sub(&self, duration: $Duration) -> Option<Self> {
+                match self.nanos.checked_sub(duration.nanos) {
+                    Some(nanos) => Some(Self { nanos }),
+                    None => None,
+                }
+            }
+
+            /// Returns the amount of time elapsed from another instant to this one.
+            /// If `other` is later that `&self` the returned value is negative.
+            ///
+            /// # Panics
+            /// Panics if an overflow in the underyling data structure happens.
+            pub fn duration_since(&self, other: Self) -> $Duration {
+                match self.nanos.checked_sub(other.nanos) {
+                    Some(nanos) => <$Duration>::from_nanos(nanos),
+                    None => panic!(),
```

Comment:
I based my instant types on `std::timer::Instant`. This type also has *checked* and *unchecked* versions of `duration_since`:
- [std::timer::Instant::duration_since](https://doc.rust-lang.org/std/time/struct.Instant.html#method.duration_since),
- [std::timer::Instant::checked_duration_since](https://doc.rust-lang.org/std/time/struct.Instant.html#method.checked_duration_since).

What's quite surprising is that `std::timer::Instant::duration_since` panics or saturates depending on the Rust version. I decided to consistently panic (instead of saturating) in all methods of my instant and duration types.

Implementing `duration_since` probably isn't necessary. I was simply implementing all methods that `std::timer::Instant` implements. So, in conclusion, my instant types are like `std::timer::Instant` with `std::chrono::type_point`'s features:
- they can have negative value,
- they implement `duration_since_epoch` (I called it like that instead of `time_since_epoch` because the rest of the methods use *duration*).